native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeTestMojo.java

```diff
@@ -181,25 +201,112 @@ public void execute() throws MojoExecutionException {
         runNativeTests(outputDirectory.toPath().resolve(NATIVE_TESTS_EXE));
     }
 
-    private void configureEnvironment() {
-        List<Plugin> plugins = new ArrayList<>();
+    protected void configureEnvironment() {
+        // set the default test classes location
+        testClassesDirectory = project.getBuild().getTestOutputDirectory();
 
         Plugin surefire = project.getPlugin("org.apache.maven.plugins:maven-surefire-plugin");
+        Plugin failsafe = project.getPlugin("org.apache.maven.plugins:maven-failsafe-plugin");
+        String currentGoal = mojoExecution.getGoal();
+        boolean testPluginProcessed = false;
+
         if (surefire != null) {
-            plugins.add(surefire);
+            testPluginProcessed = processTestPluginConfig(surefire, currentGoal);
         }
 
-        Plugin failsafe = project.getPlugin("org.apache.maven.plugins:maven-failsafe-plugin");
-        if (failsafe != null) {
-            plugins.add(failsafe);
+        if (!testPluginProcessed && failsafe != null) {
+            // Surefire was not configured with an execution for the current goal, so try Failsafe
+            testPluginProcessed = processTestPluginConfig(failsafe, currentGoal);
+        }
+
+        if (!testPluginProcessed) {
+            // neither Surefire nor Failsafe has an execution for the current goal,
+            // so use the configuration for whichever plugin's default goal matches
+            // the current goal
+            if (mojoExecution.getGoal().equals(TEST_GOAL) && surefire != null) {
+                // the current goal is "test" which is Surefire's default goal
+                processTestPluginConfig(surefire, null);
+                getLog().info("Using configuration from " + surefire.getArtifactId());
+            } else if (mojoExecution.getGoal().equals(INTEGRATION_TEST_GOAL) && failsafe != null) {
```

Comment:
Could we put this check into the `NativeIntegrationTestMojo`?

Reply:
This part of the code has been removed after I found out that plugins always have a default execution and Maven combines the global and execution configurations together in the model. So the Surefire and Failsafe plugins always  their default executions matching the `test` or `integration-test` goals. The only reason the code still looks in both plugins for the current goal is because someone _might_ bind Surefire to the `integration-test` goal or Failsafe to the `test` goal.